crates/valence_client/src/world_border.rs

```diff
@@ -0,0 +1,336 @@
+
+#[derive(SystemSet, Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub struct UpdateWorldBorderPerClientSet;
+
+pub(crate) fn build(app: &mut App) {
+    app.configure_set(
+        UpdateWorldBorderPerInstanceSet
+            .in_base_set(CoreSet::PostUpdate)
+            .before(WriteUpdatePacketsToInstancesSet),
+    )
+    .configure_set(
+        UpdateWorldBorderPerClientSet
+            .in_base_set(CoreSet::PostUpdate)
+            .before(FlushPacketsSet),
+    )
+    .add_event::<SetWorldBorderSizeEvent>()
+    .add_systems(
+        (
+            handle_wb_size_change.before(handle_diameter_change),
+            handle_diameter_change,
+            handle_lerp_transition,
+            handle_center_change,
+            handle_warn_time_change,
+            handle_warn_blocks_change,
+            handle_portal_teleport_bounary_change,
+        )
+            .in_set(UpdateWorldBorderPerInstanceSet),
+    )
+    .add_system(handle_border_for_player.in_set(UpdateWorldBorderPerClientSet));
+}
+
+/// A bundle contains necessary component to enable world border.
+/// This struct implements [`Default`] trait that returns a bundle using Minecraft Vanilla defaults.
+#[derive(Bundle)]
+pub struct WorldBorderBundle {
+    pub center: WorldBorderCenter,
+    pub diameter: WorldBorderDiameter,
+    pub portal_teleport_boundary: WorldBorderPortalTpBoundary,
+    pub warning_time: WorldBorderWarnTime,
+    pub warning_blocks: WorldBorderWarnBlocks,
+    pub moving: MovingWorldBorder,
+}
+
+impl WorldBorderBundle {
+    /// Create a new world border with specified center and diameter
+    pub fn new(center: impl Into<DVec2>, diameter: f64) -> Self {
+        Self {
+            center: WorldBorderCenter(center.into()),
+            diameter: WorldBorderDiameter(diameter),
+            portal_teleport_boundary: WorldBorderPortalTpBoundary(DEFAULT_PORTAL_LIMIT),
+            warning_time: WorldBorderWarnTime(DEFAULT_WARN_TIME),
+            warning_blocks: WorldBorderWarnBlocks(DEFAULT_WARN_BLOCKS),
+            moving: MovingWorldBorder {
+                old_diameter: diameter,
+                new_diameter: diameter,
+                speed: 0,
+                timestamp: Instant::now(),
+            },
+        }
+    }
+}
+
+impl Default for WorldBorderBundle {
+    fn default() -> Self {
+        Self::new([0.0, 0.0], DEFAULT_DIAMETER)
+    }
+}
+
+#[derive(Component)]
+pub struct WorldBorderCenter(pub DVec2);
+
+#[derive(Component)]
+pub struct WorldBorderWarnTime(pub i32);
+
+#[derive(Component)]
+pub struct WorldBorderWarnBlocks(pub i32);
+
+#[derive(Component)]
+pub struct WorldBorderPortalTpBoundary(pub i32);
+
+#[derive(Component)]
+pub struct WorldBorderDiameter(f64);
+
+impl WorldBorderDiameter {
+    pub fn diameter(&self) -> f64 {
+        self.0
+    }
+}
+
+#[derive(Component)]
+pub struct MovingWorldBorder {
+    pub old_diameter: f64,
+    pub new_diameter: f64,
+    pub speed: i64,
+    pub timestamp: Instant,
+}
```

Comment:
I'm not sure commenting on each field is necessary because this struct is essentially the `Set Border Lerp Size` packet with a timestamp. I also believe that all fields should be public if we want to keep both Event and Component Change Tracking as ways to modify world border diameter

Reply:
Sounds good, but the type still needs a brief comment to explain what it is.